runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -102,34 +104,237 @@ void J9::RecognizedCallTransformer::process_java_lang_StringUTF16_toBytes(TR::Tr
    treetop->insertAfter(TR::TreeTop::create(comp(), TR::Node::create(node, TR::treetop, 1, newCallNode)));
    }
 
-void J9::RecognizedCallTransformer::processUnsafeAtomicCall(TR::TreeTop* treetop, TR::Node* node, TR::SymbolReferenceTable::CommonNonhelperSymbol helper)
+/*
+Transform an Unsafe atomic call to diamonds with equivalent semantics
+
+                          yes
+isObjectNull ------------------------------------------>
+    |                                                  |
+    | no                                               |
+    |                     yes                          |
+isNotLowTagged ---------------------------------------->
+    |                                                  |
+    |  no                                              |
+    |           no                                     |
+isFinal ---------------------->                        |
+    |                         |                        |
+    | yes                     |                        |
+    |                         |                        |
+call the                calculate address      calculate address
+original method         for static field       for instance field
+    |                         |                and absolute address
+    |                         |                        |
+    |                         |________________________|
+    |                                     |
+    |                         xcall atomic method helper
+    |                                     |
+    |_____________________________________|
+                    |
+      program after the original call
+
+Block before the transformation: ===>
+
+start Block_A
+  ...
+xcall Unsafe.xxx
+  ...
+end Block_A
+
+Blocks after the transformation: ===>
+
+start Block_A
+...
+ifacmpeq -> <Block_E>
+  object
+  aconst null
+end Block_A
+
+start Block_B
+iflcmpne -> <Block_E>
+  lload <offset>
```

Comment:
There's an and involved here for the tag that is missing in this comment right?

Reply:
Thanks for pointing this out. I've updated the comment as well as the code such that the comparison matches the one in the comment.